moviebot/nlu/nlu.py

```diff
@@ -36,17 +33,146 @@ def __init__(self, config):
+                user_utterance, UserIntents.HI
+            )
+            or [DialogueAct(UserIntents.UNK, [])]
+        )
+
+    def _process_last_agent_dacts(
+        self, user_utterance: UserUtterance, last_agent_dacts: List[DialogueAct]
+    ) -> List[DialogueAct]:
+        """Processes response to agent dialogue acts from previous turn.
+
+        Args:
+            user_utterance: User utterance.
+            last_agent_dacts: Last agent dialogue acts.
+
+        Returns:
+            A list of dialogue acts.
+        """
+        for last_agent_dact in last_agent_dacts:
+            if last_agent_dact.intent == AgentIntents.WELCOME:
+                user_dacts = self._follow_up_welcome(user_utterance)
+                if user_dacts:
+                    return user_dacts
+            elif last_agent_dact.intent == AgentIntents.ELICIT:
+                user_dacts = self._follow_up_elicit(
+                    user_utterance, last_agent_dact
+                )
+                if user_dacts:
+                    return user_dacts
+        return []
+
+    def _process_feedback(
+        self, user_utterance: UserUtterance
+    ) -> List[DialogueAct]:
+        """Processes feedback from the user.
+
+        Args:
+            user_utterance: User utterance.
+
+        Returns:
+            A list of dialogue acts.
+        """
+        feedback_intents = [
+            self.intents_checker.check_reject_intent,
+            self.intents_checker.check_inquire_intent,
+            self.intents_checker.check_reveal_voluntary_intent,
+            self._convert_deny_to_inquire,
+        ]
+
+        for check_intent in feedback_intents:
+            user_dacts = check_intent(user_utterance)
+            if user_dacts:
+                return user_dacts
+        return []
+
+    def _follow_up_welcome(
+        self, user_utterance: UserUtterance
+    ) -> List[DialogueAct]:
+        """Follow up on welcome intent.
+
+        Args:
+            user_utterance: User utterance.
+
+        Returns:
+            A list of dialogue acts.
+        """
+        return self.intents_checker.check_reveal_voluntary_intent(
+            user_utterance
+        ) or self.intents_checker.check_basic_intent(
+            user_utterance, UserIntents.ACKNOWLEDGE
+        )
+
+    def _follow_up_elicit(
+        self, user_utterance: UserUtterance, last_agent_dact: DialogueAct
+    ) -> List[DialogueAct]:
+        """Follow up on elicit intent.
+
+        Args:
+            user_utterance: User utterance.
+            last_agent_dact: Last agent dialogue act.
+
+        Returns:
+            A list of dialogue acts.
+        """
+        user_dacts = self.intents_checker.check_reveal_intent(
+            user_utterance, last_agent_dact
+        )
+        if not user_dacts or any(
+            [
+                param.value in Values.__dict__.values()
+                for dact in user_dacts
+                for param in dact.params
+            ]
+        ):
```

Comment:
Could you move any(...) before the loop and give it an explainable name? It would make it much easier to read and understand

Reply:
Good point. Done.